proposals/p2188.md

```diff
@@ -0,0 +1,517 @@
+is the resolved type of the nested _pattern_. Any bindings within the nested
+pattern refer to portions of the corresponding storage rather than to the
+scrutinee.
+
+```
+fn F(p: i32*);
+fn G() {
+  match ((1, 2)) {
+    // `n` is a mutable `i32`.
+    case (var n: i32, 1) => { F(&n); }
+    // `n` and `m` are the elements of a mutable `(i32, i32)`.
+    case var (n: i32, m: i32) => { F(if n then &n else &m); }
+  }
+}
+```
+
+Pattern matching precedes the initialization of the storage for any `var`
+patterns. An introduced variable is only initialized if the complete pattern
+matches.
+
+```
+class X {
+  destructor { Print("Destroyed!"); }
+}
+fn F(x: X) {
+  match ((x, 1 as i32)) {
+    case (var y: X, 0) => {}
+    case (var z: X, 1) => {}
+    // Prints "Destroyed!" only once, when `z` is destroyed.
+  }
+}
+```
+
+A `var` pattern cannot be nested within another `var` pattern. The declaration
+syntax `var` _pattern_ `=` _expresson_ `;` is equivalent to `let` `var`
+_pattern_ `=` _expression_ `;`.
+
+### Tuple patterns
+
+A tuple of patterns can be used as a pattern.
+
+-   _tuple-pattern_ ::= `(` [_expression_ `,`]\* _proper-pattern_ [`,`
+    _pattern_]\* `,`? `)`
+-   _proper-pattern_ ::= _tuple-pattern_
+
+A _tuple-pattern_ containing no commas is treated as grouping parens: the
+contained _proper-pattern_ is matched directly against the scrutinee. Otherwise,
+the behavior is as follows.
+
+A tuple pattern is matched left-to-right. The scrutinee is required to be of
+tuple type.
+
+Note that a tuple pattern must contain at least one _proper-pattern_. Otherwise,
+it is a tuple-valued expression. However, a tuple pattern and a corresponding
+tuple-valued expression are matched in the same way because `==` for a tuple
+compares fields left-to-right.
+
+### Struct patterns
+
+A struct can be matched with a struct pattern.
+
+-   _proper-pattern_ ::= `{` [_field-init_ `,`]\* _proper-field-pattern_ [`,`
+    _field-pattern_]\* `}`
+-   _proper-pattern_ ::= `{` [_field-pattern_ `,`]+ `...` `}`
+-   _field-init_ ::= _designator_ `=` _expression_
+-   _proper-field-pattern_ ::= _designator_ `=` _proper-pattern_
+-   _proper-field-pattern_ ::= _binding-pattern_
+-   _field-pattern_ ::= _field-init_
+-   _field-pattern_ ::= _proper-field-pattern_
+
+A struct pattern resembles a struct literal, with at least one field initialized
+with a proper pattern:
+
+```
+match ({.a = 1, .b = 2}) {
+  // Struct literal as an expression pattern.
+  case {.b = 2, .a = 1} => {}
+  // Struct pattern.
+  case {.b = n: i32, .a = m: i32} => {}
+}
+```
+
+The scrutinee is required to be of struct type, and to have the same set of
+field names as the pattern. The pattern is matched left-to-right, meaning that
+matching is performed in the field order specified in the pattern, not in the
+field order of the scrutinee. This is consistent with the behavior of matching
+against a tuple-valued expression, where the left operand determines the order
```

Comment:
"_pattern_ operand" might be clearer; as it is, I couldn't remember which operand was on the left in the expression case.